Revise for readability (#75) (#176)
nginx/documentation: content/nginx/deployment-guides/amazon-web-services/high-availability-network-load-balancer.md

weight: 100
---

> These instructions explain how to deploy NGNX Plus in the Amazon Web Services (AWS) cloud. Use these instructions to create a highly available, active-active deployment with load balancing.

Reviewer: ```suggestion
These instructions explain how to deploy F5 NGINX Plus in the Amazon Web Services (AWS) cloud. Use these instructions to create a highly available, active-active deployment with load balancing.
```

<span id="about-nlb"></span>
## About AWS NLB

AWS NLB uses a [flow hash routing algorithm](https://docs.aws.amazon.com/elasticloadbalancing/latest/userguide/how-elastic-load-balancing-works.html#routing-algorithm). It promotes fast, efficient load balancing at the connection level (Layer 4). This means that AWS NLB is ideal for fast load balancing of TCP traffic. It can handle millions of requests per second while maintaining ultra‑low latencies. As a result, AWS NLB easily handles traffic volume patterns that suddenly and dramatically change. AWS NLB supports both static IP addresses and [Elastic IP addresses](http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/elastic-ip-addresses-eip.html). This is a change from previous AWS solutions.

<span id="about-nginx"></span>
## About NGINX Plus

[NGINX Plus](https://www.f5.com/products/nginx/nginx-plus) is the commercial version of [NGINX Open Source](http://nginx.org/en) software, which extends the power of NGINX with enterprise-ready capabilities.

> NGINX Plus complements AWS NLB. It operates at Layer 7 (the application layer) where it uses advanced criteria when load balancing. These criteria include schemes that rely on content of requests and the results of [active health checks]({{< relref "../../admin-guide/load-balancer/http-health-check.md" >}}).

Reviewer 2: ```suggestion
NGINX Plus complements AWS NLB. It operates at Layer 7 (the application layer) where it uses advanced criteria when load balancing. These criteria include schemes that rely on content of requests and the results of [active health checks]({{< ref "/nginx/admin-guide/load-balancer/http-health-check.md" >}}).
```

In this step, you create a _target group_, using the **Step 2: Configure Routing** window. The target group contains the set of EC2 instances across which your AWS NLB load balances traffic. You specify those EC2 instances later, in the step [Register Instances in the Target Group](#nlb-register-instances)).

1. In the **Target group** section, use the following values to create the target group:

    - **Target group** – Select <span style="color:#666666; font-weight:bolder">New target group</span>
    - **Name** – Enter the name of the new target group. For these instructions use <span style="color:#666666; font-weight:bolder">aws-nlb-tg</span>.
    - **Protocol** – Select <span style="color:#666666; font-weight:bolder">TCP</span> (the only available option)
    - **Port** – Enter <span style="color:#666666; font-weight:bolder">80</span>. This is the same value you specified for the **Load Balancer Port** field in Step 3 of the [previous section](#nlb-create).
    - **Target type** – Enter <span style="color:#666666; font-weight:bolder">instance</span>

2. In the **Health checks** section, open the **Advanced health check settings** subsection and enter the following values:

>     - **Protocol** – Enter either <span style="color:#666666; font-weight:bolder">TCP</span>, <span style="color:#666666; font-weight:bolder">HTTP</span>, or <span style="color:#666666; font-weight:bolder">HTTPS</span>, as described below. This field specifies the protocol the AWS NLB uses when sending health checks.

Reviewer: ```suggestion
    - **Protocol** – Enter either *HTTP*, or *HTTPS*, as described below. This field specifies the protocol the AWS NLB uses when sending health checks.
```

    - **TCP** - AWS NLB attempts to open a TCP connection to send a health check. The port it uses is specified in the next field. If you select TCP you must also define **Port**.

>     - **Port** – Select <span style="color:#666666; font-weight:bolder">traffic port</span>. This is the port on the target instances to which the AWS NLB sends health checks. These instructions use **traffic port** to send health checks to the same port as regular traffic. This value is required if the value for **Protocol** is **TCP**.

Reviewer: ```suggestion
    - **Port** – Select *traffic port*. This is the port on the target instances to which the AWS NLB sends health checks. These instructions use **traffic port** to send health checks to the same port as regular traffic. This value is required if the value for **Protocol** is **TCP**.
```

3. Select the <span style="background-color:#cccccc; font-weight:bolder; white-space: nowrap;"> Next: Register Targets </span> button. The **Step 3: Register Targets** window opens.

<span id="nlb-register-instances"></span>
### Register Instances in the Target Group

In this step, you add instances to the empty target group you created in the previous section. Use the the **Step 3: Register Targets** window to add both NGINX Plus load balancer instances.

1. In the **Instances** table, select the buttons for both NGINX Plus load balancer instances. Select the instance names <span style="color:#666666; font-weight:bolder">ngx-plus-1</span> and <span style="color:#666666; font-weight:bolder">nginx-plus-2</span>.

   <a href="/nginx/images/aws-nlb-load-balancer-register-targets.png"><img src="/nginx/images/aws-nlb-load-balancer-register-targets.png" alt="" width="1024" height="847" class="aligncenter size-full wp-image-54846" style="border:2px solid #666666; padding:2px; margin:2px;" /></a>

> 2. Select the <span style="background-color:#3366cc; color:white; white-space: nowrap;"> Add to registered </span> button above the table. The instances are added to the **Registered targets** table.

Reviewer: ```suggestion
2. Select the *Add to registered* button above the table. The instances are added to the **Registered targets** table.
```

   <a href="/nginx/images/aws-nlb-load-balancer-targets-display.png"><img src="/nginx/images/aws-nlb-load-balancer-targets-display.png" alt="" width="1024" height="874" class="aligncenter size-full wp-image-54845" style="border:2px solid #666666; padding:2px; margin:2px;" /></a>

> 3. Select the <span style="background-color:#cccccc; font-weight:bolder; white-space: nowrap;"> Next: Review </span> button. The **Step 4: Review** window opens.

Reviewer: ```suggestion
3. Select the *Next: Review* button. The **Step 4: Review** window opens.
```

<span id="nlb-launch"></span>
### Launch the AWS NLB

In this step you use the **Step 4: Review** window to verify settings and launch AWS NLB.

1. In the **Step 4: Review** window verify that the settings are correct.

>      - If so, select the <span style="background-color:#3366cc; color:white; white-space: nowrap;"> Create </span> button.
>      - If the settings are not correct, select the <span style="background-color:#cccccc; font-weight:bolder; white-space: nowrap;"> Previous </span> button to go back to a previous screen. Make required change(s) on previous screen(s). Then, return to the **Step 4: Review** window to select <span style="background-color:#3366cc; color:white; white-space: nowrap;"> Create </span>.

Reviewer: ```suggestion
     - If so, select the <span style="background-color:#3366cc; color:white; white-space: nowrap;"> Create </span> button.
     - If the settings are not correct, select the *Previous* button to go back to a previous screen. Make required change(s) on previous screen(s). Then, return to the **Step 4: Review** window to select *Create*.
```

   <a href="/nginx/images/aws-nlb-load-balancer-review.png"><img src="/nginx/images/aws-nlb-load-balancer-review.png" alt="" width="1023" height="789" class="aligncenter size-full wp-image-54843" style="border:2px solid #666666; padding:2px; margin:2px;" /></a>

>     The <span style="background-color:#3366cc; color:white; white-space: nowrap;"> Create </span> action provisions the AWS NLB. A success message displays when the provisioning operation finishes. It can take a few minutes to provision the AWS NLB.

Reviewer: ```suggestion
    The *Create* action provisions the AWS NLB. A success message displays when the provisioning operation finishes. It can take a few minutes to provision the AWS NLB.
```

2. Select the <span style="background-color:#3366cc; color:white; white-space: nowrap;"> Close </span> button to return to the **Load Balancers** dashboard.

   <a href="/nginx/images/aws-nlb-load-balancer-success.png"><img src="/nginx/images/aws-nlb-load-balancer-success.png" alt="" width="1024" height="345" class="aligncenter size-full wp-image-54842" style="border:2px solid #666666; padding:2px; margin:2px;" /></a>

> 3. Observe the values in the **State** column of the table. When a value changes to <span style="color:#666666; font-weight:bolder">active</span>, you can display details about the provisioned AWS NLB. Select the button in an AWS NLB row to display its details.

Reviewer: ```suggestion
3. Observe the values in the **State** column of the table. When a value changes to *active*, you can display details about the provisioned AWS NLB. Select the button in an AWS NLB row to display its details.
```

   
   <a href="/nginx/images/aws-nlb-load-balancer-active-state.png"><img src="/nginx/images/aws-nlb-load-balancer-active-state.png" alt="" width="1024" height="634" class="aligncenter size-full wp-image-54841" style="border:2px solid #666666; padding:2px; margin:2px;" /></a>

> 4. Next, verify that the AWS NLB is working. Open a new browser window and navigate to the AWS NLB's public DNS name. You can find the **DNS name** field in the **Basic Configuration** section of the **Load Balancers** dashboard. **Note:** If you copy and paste the DNS name, be sure to exclude the parenthesized words at the end, for example, <span style="color:#666666; font-weight:bolder">(A Record)</span>.

Reviewer: ```suggestion
4. Next, verify that the AWS NLB is working. Open a new browser window and navigate to the AWS NLB's public DNS name. You can find the **DNS name** field in the **Basic Configuration** section of the **Load Balancers** dashboard. **Note:** If you copy and paste the DNS name, be sure to exclude the parenthesized words at the end, for example, *(A Record)*.
```

    If you see the **Welcome to nginx!** page then the AWS NLB successfully forwarded a request to one of the two NGINX Plus instances.

   <a href="/nginx/images/aws-nlb-welcome-to-nginx.png"><img src="/nginx/images/aws-nlb-welcome-to-nginx.png" alt="" width="1023" height="328" class="aligncenter size-full wp-image-54840" style="border:2px solid #666666; padding:2px; margin:2px;" /></a>

5. Last, verify that the NGINX Plus load balancer can reach backend applications.

   - Append <span style="white-space: nowrap; font-weight:bold;">/backend-one</span> to the public DNS name in the open browser window. If <span style="color:#666666; font-weight:bolder">App 1</span> is displayed then you have reached the **App 1** NGINX backend instance.

   <a href="/nginx/images/aws-nlb-app1.png"><img src="/nginx/images/aws-nlb-app1.png" alt="" width="1024" height="491" class="aligncenter size-full wp-image-54839" style="border:2px solid #666666; padding:2px; margin:2px;" /></a>

>     - Append <span style="white-space: nowrap; font-weight:bold;">/backend-two</span> to the public DNS name in the open browser window. If <span style="color:#666666; font-weight:bolder">App 2</span> is displayed then you have reached the **App 2** NGINX backend instance.

Reviewer: ```suggestion
    - Append */backend-two* to the public DNS name in the open browser window. If *App 2* is displayed then you have reached the **App 2** NGINX backend instance.
```

You can either complete these individual steps:

- [Create EC2 Instances and Install the NGINX Software](#create-instance-install-nginx)
- [Configure NGINX Open Source on the Web Servers](#configure-web-servers)
- [Configure NGINX Plus on the Load Balancers](#configure-load-balancers)

OR

Use our Packer and Terraform scripts to completely automate the process:

- [Automate Instance Setup with Packer and Terraform](#create-instances-automated)

Once you have created and configured the EC2 instances, your prerequisites are complete. Continue to [Configure an AWS Network Load Balancer](#nlb-configure).

<span id="create-instance-install-nginx"></span>
#### Create EC2 Instances and Install the NGINX Software

The deployed solution in these instructions uses six EC2 instances. Two instances run NGINX Plus. These load balance traffic to the other four instances, which run NGINX Open Source as a web server. The four NGINX Open Source instances deploy in two pairs; each pair runs a different app.

> <span style="white-space: nowrap;">Step‑by‑step</span> instructions for creating EC2 instances and installing NGINX Open Source and NGINX Plus are available. Refer to our deployment guide, [Creating Amazon EC2 Instances for NGINX Open Source and NGINX Plus]({{< relref "ec2-instances-for-nginx.md" >}}).

Reviewer: ```suggestion
*Step‑by‑step* instructions for creating EC2 instances and installing NGINX Open Source and NGINX Plus are available. Refer to our deployment guide, [Creating Amazon EC2 Instances for NGINX Open Source and NGINX Plus]({{< ref "/nginx/deployment-guides/amazon-web-services/ec2-instances-for-nginx.md" >}}).
```

We use absolute paths for everything anyway: I'm slowly swapping `relref` to `ref` everywhere to reflect this fact.

> **Note:** When installing NGINX Open Source or NGINX Plus, you connect to each instance over SSH. To save time, leave the SSH connection to each instance open after installing the software. This way, you can reuse the connection when configuring the instance.

Reviewer: ```suggestion
{{< note >}} When installing NGINX Open Source or NGINX Plus, you connect to each instance over SSH. To save time, leave the SSH connection to each instance open after installing the software. This way, you can reuse the connection when configuring the instance. {{< note >}} 
```

<a href="/nginx/images/aws-nlb-instances-summary.png"><img src="/nginx/images/aws-nlb-instances-summary.png" alt="" width="1024" height="263" class="aligncenter size-full wp-image-54856" style="border:2px solid #666666; padding:2px; margin:2px;" /></a>

<span id="configure-web-servers"></span>
#### Configure NGINX Open Source on the Web Servers

> Configure NGINX Open Source instances as web servers. These should return a page specifying the server name, address, and other information. As an example, here's the page returned by <span style="color:#666666; font-weight:bolder; white-space: nowrap;">App 1</span>:

Reviewer: ```suggestion
Configure NGINX Open Source instances as web servers. These should return a page specifying the server name, address, and other information. As an example, here's the page returned by *App 1*:
```

   <a href="/nginx/images/aws-nlb-app1.png"><img src="/nginx/images/aws-nlb-app1.png" alt="" width="1024" height="491" class="aligncenter size-full wp-image-54839" style="border:2px solid #666666; padding:2px; margin:2px;" /></a>

> Use the <span style="white-space: nowrap;">Step‑by‑step</span> instructions in our deployment guide, <a href="../../setting-up-nginx-demo-environment#nginx-oss">Setting Up an NGINX Demo Environment</a>.

Reviewer: ```suggestion
Use the *Step‑by‑step* instructions in our deployment guide, [Setting Up an NGINX Demo Environment]({{< ref "/nginx/deployment-guides/setting-up-nginx-demo-environment.md" >}}).
```

  - <span style="color:#666666; font-weight:bolder">ngx-oss-app2-2</span>

<span id="configure-load-balancers"></span>
#### Configure NGINX Plus on the Load Balancers

Configure NGINX Plus instances as load balancers. These distribute requests to NGINX Open Source web servers set up in [Configure NGINX Open Source on the Web Servers](#configure-web-servers).

> Use the <span style="white-space: nowrap;">Step‑by‑step</span> instructions in our deployment guide, <a href="../../setting-up-nginx-demo-environment#nginx-plus">Setting Up an NGINX Demo Environment</a>.

Reviewer: ```suggestion
Use the *Step‑by‑step* instructions in our deployment guide, [Setting Up an NGINX Demo Environment]({{< ref "/nginx/deployment-guides/setting-up-nginx-demo-environment.md" >}}).
```

Repeat the instructions on both <span style="color:#666666; font-weight:bolder; white-space: nowrap;">ngx-plus-1</span> and <span style="color:#666666; font-weight:bolder; white-space: nowrap;">ngx-plus-2</span>.

<span id="create-instances-automated"></span>
### Automate Instance Setup with Packer and Terraform

You can automate set up of the six instances described in these instructions. Automation is an alternative to creating and configuring each instance one at a time. To automate the set up, use the Packer and Terraform scripts from our [GitHub repository](https://github.com/nginxinc/NGINX-Demos/tree/master/aws-nlb-ha-asg). These scripts will:

- Configure two load balancer instances running NGINX Plus 
- Configure four web server instances running NGINX Open Source

These scripts also create a new set of networking rules and security group settings. These rules and settings help avoid conflicts with any pre‑existing network settings. After you run the scripts, continue to the [instructions for creating an AWS NLB](#nlb-configure). No further setup is required.

> **Note:** These scripts also create a new VPC. They do not use the default VPC described in the [instructions in our Deployment Guide]({{< relref "ec2-instances-for-nginx.md" >}}).

Reviewer: ```suggestion
{{< note >}} These scripts also create a new VPC. They do not use the default VPC described in the [instructions in our Deployment Guide]({{< ref "/nginx/deployment-guides/amazon-web-services/ec2-instances-for-nginx.md >}}).  {{< /note >}}
```

To run the scripts, follow these instructions:

1. Install [Packer](https://www.packer.io/intro/getting-started/install.html) and [Terraform](https://learn.hashicorp.com/terraform).

2. Clone or download the scripts from our [GitHub repository](https://github.com/nginxinc/NGINX-Demos/tree/master/aws-nlb-ha-asg):

>    - Use the scripts in <span style="white-space: nowrap; font-weight:bold;">packer/ngx-oss</span> to create an Ubuntu AMI running NGINX Open Source.
>    - Use the scripts in <span style="white-space: nowrap; font-weight:bold;">packer/ngx-plus</span> to create an AWS Linux AMI running NGINX Plus.

Reviewer: ```suggestion
   - Use the scripts in **packer/ngx-oss** to create an Ubuntu AMI running NGINX Open Source.
   - Use the scripts in **packer/ngx-plus** to create an AWS Linux AMI running NGINX Plus.
```
